### PR TITLE
Bottilabo layout

### PR DESCRIPTION
**Pull request by bottilabo**
Added bottilabo layout that enables to change keybord layout dynamically and has language specific input mode.

## Description
Added bottilabo layout that enables to change QWERTY keybord layout dynamically into other colemak,dvorak..etc. and has language specific input mode(for now,some Japanese KANA and Romaji input mode using simultaneous keying or pre shift or post shift).

My compile option for test is
`make KEYBOARD:bottilabo:avrdude EXTRAFLAGS='-DNO_KANA -DKBLAYOUT1=KB_COLEMAK -DKBLAYOUT2=KB_DVORAK -DKBLAYOUT3=KB_ARENSITO -DKBIM1=IM_TRON'`

I've confirmed all compiled firmware on all keyboards.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by skullydazed** [comment]
Thanks for your contribution! It's always nice to see people put this much work and passion into QMK.

There are two features you may not be aware of which would help to simplify your PR a lot:

* https://docs.qmk.fm/#/feature_userspace
* https://docs.qmk.fm/#/feature_layouts

A lot of the duplication in this PR could be eliminated by making use of those.

Are the keyboards in this repo all keyboards that you own? We are concerned about the size of this PR and how much maintenance burden it may add on an ongoing basis.

**Comment by bottilabo** [comment]
Thank you for reviewing. This PR is huge, your concern is reasonable. I thought all keyboard are using Bottilabo userspace and only Japanese might test language specific IM, so I thought this PR can't divide into some PRs.

I tried to eliminate duplication. Did I get your intension?
And I've tested all of them again on all keyboards.
This is my compile option passed.
```
make all:bottilabo
make all:bottilabo:avrdude EXTRAFLAGS='-DNO_KANA -DKBLAYOUT1=KB_COLEMAK -DKBLAYOUT2=KB_DVORAK -DKBLAYOUT3=KB_ARENSITO -DKBIM1=IM_TRON'
```

As I addicted to keyboards, I own all keyboards in PR and I'm always changing them every few hours.... This is why I made library in user space. Previously, when I made some change in layout, I had to modify each keymap.c I use. But now, I only to change bottilabo/KBD_HARMONIZE and just compile!

In fact, I have more keybords and there was more language specific IM in Bottilabo userspase. But I've removed them from PR because I don't use them on a daijy basis.

File: keyboards/crkbd/keymaps/bottilabo/keymap.c [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+extern uint8_t is_master;
+
+
+int RGB_current_mode;
+
+void persistent_default_layer_set(uint16_t default_layer) {
+  eeconfig_update_default_layer(default_layer);
+  default_layer_set(default_layer);
+}
+
```
Comment (drashna) [inline keyboards/crkbd/keymaps/bottilabo/keymap.c]:
```suggestion
```

File: keyboards/crkbd/keymaps/bottilabo/keymap.c [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+#include "harmonize.h"
+
+
+
+
+#ifdef RGBLIGHT_ENABLE
+//Following line allows macro to read current RGB settings
+extern rgblight_config_t rgblight_config;
+#endif
```
Comment (drashna) [inline keyboards/crkbd/keymaps/bottilabo/keymap.c]:
```suggestion
```

File: keyboards/crkbd/keymaps/bottilabo/keymap.c [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+  if (IS_LAYER_ON(layer1) && IS_LAYER_ON(layer2)) {
+    layer_on(layer3);
+  } else {
+    layer_off(layer3);
+  }
+}
+
+void matrix_init_user(void) {
+    #ifdef RGBLIGHT_ENABLE
+      RGB_current_mode = rgblight_config.mode;
```
Comment (drashna) [inline keyboards/crkbd/keymaps/bottilabo/keymap.c]:
```suggestion
      RGB_current_mode = rgblight_get_mode();
```

File: keyboards/crkbd/keymaps/bottilabo/keymap.c [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    // set_timelog();
+  }
+
+  switch (keycode) {
+    case RGB_MOD:
+      #ifdef RGBLIGHT_ENABLE
+        if (record->event.pressed) {
+          rgblight_mode(RGB_current_mode);
+          rgblight_step();
+          RGB_current_mode = rgblight_config.mode;
```
Comment (drashna) [inline keyboards/crkbd/keymaps/bottilabo/keymap.c]:
```suggestion
          RGB_current_mode = rgblight_get_mode();
```

File: keyboards/crkbd/keymaps/bottilabo/keymap.c [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        }
+      #endif
+      return false;
+    case RGBRST:
+      #ifdef RGBLIGHT_ENABLE
+        if (record->event.pressed) {
+          eeconfig_update_rgblight_default();
+          rgblight_enable();
+          RGB_current_mode = rgblight_config.mode;
```
Comment (drashna) [inline keyboards/crkbd/keymaps/bottilabo/keymap.c]:
```suggestion
          RGB_current_mode = rgblight_get_mode();
```

File: keyboards/ergo42/keymaps/bottilabo/keymap.c [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+ */
+#include QMK_KEYBOARD_H
+#include "bootloader.h"
+#ifdef PROTOCOL_LUFA
+#include "lufa.h"
+#include "split_util.h"
+#endif
+
+
+extern keymap_config_t keymap_config;
```
Comment (drashna) [inline keyboards/ergo42/keymaps/bottilabo/keymap.c]:
```suggestion
```

File: keyboards/claw44/keymaps/bottilabo/keymap.c [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+#include QMK_KEYBOARD_H
+#ifdef PROTOCOL_LUFA
+  #include "lufa.h"
+  #include "split_util.h"
+#endif
+#ifdef SSD1306OLED
+  #include "ssd1306.h"
+#endif
+
+extern keymap_config_t keymap_config;
```
[the remark this answers is not in the document]
Comment (bottilabo) [reply keyboards/claw44/keymaps/bottilabo/keymap.c]: Thank you for very kind advice. I am currently working on the correction according to the advice!

**Comment by bottilabo** [comment]
Thaks for reviewing.
I eliminated all depricated codes and unnecessary codes, duplicated codes by plugin like technique.
And I tested all of keyboards and all of IM, keyboard layouts.

This is a build option examples below.
```
make silverbullet44:bottilabo EXTRAFLAGS='-DKBLAYOUT1=KB_COLEMAK -DNO_KANA -DKBIM1=IM_TRON -DPLUG1=PLUG_RGBLIGHT -DRGBLIGHT_ENABLE'
make naked64:bottilabo EXTRAFLAGS='-DKBLAYOUT1=KB_COLEMAK -DNO_KANA -DKBIM1=IM_TRON -DPLUG1=PLUG_RGBLIGHT -DRGBLIGHT_ENABLE'

make crkbd:bottilabo EXTRAFLAGS='-DNO_KANA -DKBLAYOUT1=KB_COLEMAK -DKBLAYOUT2=KB_DVORAK -DKBIM1=IM_NICOLA -DPLUG1=PLUG_OLED'

make claw44:bottilabo EXTRAFLAGS='-DNO_KANA -DKBLAYOUT1=KB_COLEMAK -DKBLAYOUT2=KB_DVORAK -DKBIM1=IM_NICOLA -DPLUG1=PLUG_OLED'


make naked64:bottilabo EXTRAFLAGS='-DNO_KANA -DKBLAYOUT1=KB_COLEMAK -DKBLAYOUT2=KB_DVORAK -DKBIM1=IM_NICOLA -DPLUG1=PLUG_RGBLIGHT'
make silverbullet44:bottilabo EXTRAFLAGS='-DNO_KANA -DKBLAYOUT1=KB_COLEMAK -DKBLAYOUT2=KB_DVORAK -DKBIM1=IM_NICOLA -DPLUG1=PLUG_RGBLIGHT'

make mint60:bottilabo EXTRAFLAGS='-DNO_KANA -DKBLAYOUT1=KB_COLEMAK -DKBLAYOUT2=KB_DVORAK -DKBIM1=IM_NICOLA -DPLUG1=PLUG_RGBLIGHT'

make ergo42:bottilabo EXTRAFLAGS='-DNO_KANA -DKBLAYOUT1=KB_COLEMAK -DKBLAYOUT2=KB_DVORAK -DKBIM1=IM_NICOLA'
make numatreus:bottilabo EXTRAFLAGS='-DNO_KANA -DKBLAYOUT1=KB_COLEMAK -DKBLAYOUT2=KB_DVORAK -DKBIM1=IM_NICOLA'
make treadstone48:bottilabo EXTRAFLAGS='-DNO_KANA -DKBLAYOUT1=KB_COLEMAK -DKBLAYOUT2=KB_DVORAK -DKBIM1=IM_NICOLA'
make atreus62:bottilabo EXTRAFLAGS='-DNO_KANA -DKBLAYOUT1=KB_COLEMAK -DKBLAYOUT2=KB_DVORAK -DKBIM1=IM_NICOLA'
make ergodash/mini:bottilabo EXTRAFLAGS='-DNO_KANA -DKBLAYOUT1=KB_COLEMAK -DKBLAYOUT2=KB_DVORAK -DKBIM1=IM_NICOLA'
make projectkb/alice:bottilabo EXTRAFLAGS='-DNO_KANA -DKBLAYOUT1=KB_COLEMAK -DKBLAYOUT2=KB_DVORAK -DKBIM1=IM_NICOLA'
make reviung41:bottilabo EXTRAFLAGS='-DNO_KANA -DKBLAYOUT1=KB_COLEMAK -DKBLAYOUT2=KB_DVORAK -DKBIM1=IM_NICOLA'

```

Thank you!

**Comment by bottilabo** [comment]
Thank you for reviewing.
I made a big version up.
So, I will make another pull request.

**Pull request closed without merging**